### ISSUE TITLE
Stream can be better

### ISSUE THREAD

**Issue by pH200** [staltz/cycle#108]
Let's start talking about Stream. This is long so TL;DR at the bottom.

First of all, why replicateAllInteraction$? You know it's unnecessary if everything was implemented just right. Because Stream can just be, like you said, InjectableObservable, without doing magic tricks just for interaction$. You want arguments for definitionFn observables. So you can replace them with subjects, except for interaction$ which is not actually an Observable. So you made InputProxy just for interaction$. No, it didn't have to be this way. You can inject normal objects and no tricks for interaction$. I'll show you later.

Second, sorry I was wrong about the timing that the observable returned by definitionFn gets subscribed. But, it's still not good the fact it gets subscribed by injection. Why inject() has to evaluate the observable if we can make it not to do so, like so many other [operators](https://github.com/Reactive-Extensions/RxJS/blob/master/doc/api/core/observable.md) by default?

And that brings up another bad thing about the Stream right now: It appends the dispose method to the observable(returned by definitionFn). And as soon as you called inject, you were responsible to dispose that Stream(InjectableObservable). Because it fires up a subscription. This is bad. For example:

```js
var injectableObservable = createStream(_interval => _interval);
var interval = Rx.Observable.interval(500);

// This subscription will end eventually.
// You are free whether to dispose it or not.
var subscription = injectableObservable
  .take(3)
  .map(x => x * x)
  .subscribe(next => console.log(next));

// However, the hidden subscription which spawned by
// the injection would never stop.
// The interval will keep popping new value. It never ends.
injectableObservable.inject(interval);

// Unless you call dispose. This is just weird.
// Because you normally don't call dispose on an Observable.
// And this proves inject() is not the only point where mutation
// happens, if you don't forget to call dispose.
injectableObservable.dispose();
```

In Rx library, the only derivation of Observable which has dispose() in it is called Subject. However, it never fires up a new subscription when you call onNext from it. Again, it can be better.

There's more I want to say about the implementation of render and custom-elements. But for now I'll focus on the Stream issue.

So finally, this is my attempt at making Stream better. It doesn't fire up a new subscription when you called inject. It doesn't distinguish between the normal value and interaction$. It can inject normal objects not only just Observables and interaction$. Please go take a look, I'm very **confident** it will blow your mind.

https://github.com/pH200/cycle/blob/master/src/stream.js

But, is this compatible with the current Stream and tests? **Yes**, all tests passed. And this time I modified **none of them** ([by using stream2](https://github.com/pH200/cycle/blob/bbb8d7d2d96f5f3ea2563ae3b2c57b7bf787e98e/src/stream2.js)), not even a single line. Except for only one which can be skipped because definitionFn is now call-by-need.

Finally, is it possible to make Cycle compatible with this Stream? Well, I've already made one.

https://github.com/pH200/cycle

TL;DR [Stream can be better](https://github.com/pH200/cycle/blob/bbb8d7d2d96f5f3ea2563ae3b2c57b7bf787e98e/src/stream.js)

**Comment by staltz** [comment]
Thanks for the challenging comments, I really appreciate it and I'll seriously consider changing the implementation.

> Why inject() has to evaluate the observable if we can make it not to do so, like so many other operators by default?

I'd rather avoid introducing new concepts such as `reactiveNode` and `connect()`. I prefer to have just `inject()` as the only imperative call that should exist in a Cycle program. If `inject` would not evaluate the observable, then nothing would happen. At some point we need to `subscribe()` so that the program can start working. Because I want to have only `inject`, then we must do the `subscribe` inside `inject`.

> But, is this compatible with the current Stream and tests? Yes, all tests passed. 

It's a mistake that the current tests in Cycle have no test case for circular dependency injection. Cycle Groups repo [has one](https://github.com/erykpiast/cyclejs-group/blob/1424669059d4a1338577854c5d652f567fad70ad/test/create-streams-group.spec.js#L96), and I should do this type of test also in Cycle core. When that's there, I'm quite sure your Stream proposal will not pass the test, since inject has to subscribe.

**Comment by pH200** [comment]
> ... Cycle Groups repo [has one](https://github.com/erykpiast/cyclejs-group/blob/1424669059d4a1338577854c5d652f567fad70ad/test/create-streams-group.spec.js#L96),

Before I get started ranting about the necessity of explicit `subscribe()`. I would like to mention that my Stream proposal has no problem passing this test. And again, no modification was made for this test. Because in this test we still have `subscribe`. And it shows my Stream can handle the circular injection, too.

https://github.com/pH200/cyclejs-group

However, you're right, if there's no `subscribe()`, nothing would happen. And that's exactly what I want in the first place. I'll explain this later.

**Comment by staltz** [comment]
By the way, you gave me some ideas, and I'm experimenting how Cycle would be without Streams. My current guess is to embed the user function into the framework, so the developer doesn't need to explicitly code it. The developer just needs to write the `computer()` function, which takes `interaction$` as input and outputs `vtree$`, then you give this function to some imperative call, like `applyToDOM(computer, '.js-container')`.

So hold your ranting for a while, let's see if I can make this better.

**Comment by pH200** [comment]
**SPOILER ALERT**

[Does this help?](https://github.com/pH200/cycle/blob/master/examples/adapter/hello.js)

**Comment by staltz** [comment]
Yes, that is quite close to what I'm doing, but at the same time removing Stream. 
If I make a branch called `nostream`, would you send a PR? I'd be glad to see you on the contributor list. ;)

**Comment by pH200** [comment]
I would love to do that! :)

**Comment by staltz** [comment]
Alright, branch `nostream` exists now.

**Comment by pH200** [comment]
Yeah, I'm doing work. However, I'm not going to make it API compatible right now. We can discuss this when we're going to review.

**Comment by staltz** [comment]
It doesn't need to be compatible. I will most likely do some work on `nostream` branch before merging to master.

**Comment by staltz** [comment]
Todo list:
- [x] Rewrite: remove Stream, make user() function implicit in the framework
- [x] Rework tests and examples
- [ ] Add test for #110
- [ ] Refactor src folder structure
- [ ] Refactor src/rendering/render-dom.js
- [ ] README
- [ ] Documentation
- [ ] TypeScript definition

**Comment by staltz** [comment]
v0.21.0 solves this issue
